### PR TITLE
feat: Forward refs for input components

### LINKED ISSUE #1050: Forward refs for Inputs
- Move **InputRef > Ref**
- Implement Code mod

Target Components
- Text Input
- Text Area
- Radio
- Checkbox
- Switch
- Select One in Labs

### PR DESCRIPTION
**Pull request by jamesfan**
## Summary

Fixes #1050.

Updates the following input components (using `createComponent`) to forward refs using the `ref` attribute instead of the `inputRef` prop:

- Checkbox
- Color Input
  - Not in the original scope, but I updated it since it wraps Text Input
- Color Preview
  - Not in the original scope, but I updated it since it wraps Color Input
- Radio
- Select (Main)
  - Not in the original scope, but I updated it anyway it to maintain parity with the rest of the input components in the Main package
  - Didn't support `inputRef` to begin with
- Select (Preview/Labs)
  - Didn't support `inputRef` to begin with
- Switch
- Text Input
- Text Area

Also updates Combobox and Search Bar to work with the updated Text Input (Combobox uses Text Input and Search Bar uses Combobox).

## Checklist

- [x] Implement codemod for converting `inputRef` to `ref`
- [x] Update component docs to replace `inputRef` with `ref`
- [x] Update 5.0 Migration Guide
- [x] Update unit tests to use `verifyComponent` to test ref forwarding (among other things)

### CODE REVIEW

File: modules/labs-react/combobox/lib/Combobox.tsx
Comment (jamesfan) [inline modules/labs-react/combobox/lib/Combobox.tsx:199]: Calling out this funky bit of casting I had to do to get TS to stop complaining about the lack of a `ref` property.

@NicholasBoll We discussed this code a bit earlier. I ended up not using `PropsWithAs` in favor of adding `ref` to the casted interface explicitly.

File: modules/labs-react/combobox/lib/Combobox.tsx
Comment (jamesfan) [inline modules/labs-react/combobox/lib/Combobox.tsx:421]: I had to modify this type definition to get TS to stop complaining about the `ref` below. I don't think this definition is necessary (TS seems to be able to infer just fine without it), however, so I may remove it.

File: modules/labs-react/select/spec/Select.spec.tsx
Comment (jamesfan) [inline modules/labs-react/select/spec/Select.spec.tsx:61]: The Labs Select doesn't handle a consumer-provided onKeyDown handler correctly, so it fails the callback test in `verifyComponent`. Addressing this in a later PR and testing ref forwarding separately in the meantime.

**Review by jamesfan — COMMENTED** [review]
I haven't implemented the codemod yet, but everything else in this PR is ready for review.

@NicholasBoll I haven't updated the unit tests to use `screen` yet. Will look into this later (or save it for a later PR).

File: modules/labs-react/combobox/spec/Combobox.spec.tsx
Comment (NicholasBoll) [inline modules/labs-react/combobox/spec/Combobox.spec.tsx:102]: The rest of this file uses the `screen` interface. Also, why did you need to await? Would the following work?
```suggestion
    it('should set the ref to the child input element', () => {
      const ref: React.RefObject<HTMLInputElement> = React.createRef();
      render(
        <Combobox>
          <TextInput ref={ref} />
        </Combobox>
      );
      expect(screen.getByRole('combobox')).toEqual(ref.current);
    });
```

File: modules/labs-react/select/spec/Select.spec.tsx
Comment (NicholasBoll) [reply modules/labs-react/select/spec/Select.spec.tsx:61]: This could probably be done pretty easily. Without using `mergeProps`, you instead do something like the following:

```tsx
const { onKeyDown } = props

const handlers = {
  onKeyDown(event) {
    onKeyDown?.(event)
    // other stuff
  }
}
```

**Review by NicholasBoll — APPROVED** [review]

File: modules/labs-react/combobox/spec/Combobox.spec.tsx
Comment (jamesfan) [reply modules/labs-react/combobox/spec/Combobox.spec.tsx]: @NicholasBoll The `async` and `await` are there from me copy-and-pasting [an old input ref test](https://github.com/Workday/canvas-kit/pull/1068/files#diff-00d14d17114b0b458110cb29abeca60151e11e4ff55e6b31522f84eeb9b9c688L58) and using it here. I thought we might be using `findByRole` due to `ref.current` not being set immediately, but the test seems to work fine if I remove `async/await` and use `getByRole` instead.

Regardless, I updated the test to use `screen`. Instead of importing `screen`, the Combobox tests set `screen` to the return value of `render` and run queries against it -- your code above works as long as I set `const screen = render(<Combobox>...</Combobox>)`.

File: modules/react/color-picker/lib/ColorInput.tsx
Comment (alanbsmith) [inline modules/react/color-picker/lib/ColorInput.tsx:151]: Comment (not an RC) I _think_ `createComponent` will handle merging this `onChange` for you. I'd need to poke at the code a bit to be sure though. Maybe something to look at. Your call.

File: modules/react/color-picker/spec/ColorInput.spec.tsx
Comment (alanbsmith) [inline modules/react/color-picker/spec/ColorInput.spec.tsx:16]: This is a nice test clean up. 👍🏻

File: modules/react/select/README.md
Comment (alanbsmith) [inline modules/react/select/README.md:19]: 👏🏻 Good call out here.

**Review by alanbsmith — APPROVED** [review]
Looks great! I had one question, but nothing to hold this PR up. Thanks for the docs improvements as well. 👏🏻

File: modules/docs/mdx/5.0-MIGRATION-GUIDE.mdx
Comment (jamesfan) [reply modules/docs/mdx/5.0-MIGRATION-GUIDE.mdx:219]: Thanks!

File: modules/labs-react/pagination/lib/Pagination/GoTo/TextInput.tsx
Comment (jamesfan) [reply modules/labs-react/pagination/lib/Pagination/GoTo/TextInput.tsx:7]: It's related to the changes in this PR, specifically the use of `createComponent` for `TextInput`. If you look at [the changes to `TextInputProps`](https://github.com/Workday/canvas-kit/pull/1068/files#diff-de6c3686ab8bc1cfd55f609d7988122166a7249cc92fa00829fc91668d2243c1R13) in this PR, you'll see that it no longer extends from `React.InputHTMLAttributes<HTMLInputElement>`. This is because that portion of the `TextInput` interface (the HTML attributes of the underlying element being represented by `TextInput`) is now being handled by `createComponent` rather than being specified in the `TextInputProps` definition.

Without `value` in `GoToTextInputProps`, TS complains about `value` not existing in `GoToTextInputProps` when it's referenced later in this file (because `value` is no longer in `TextInputProps`).

File: modules/labs-react/combobox/spec/Combobox.spec.tsx
Comment (NicholasBoll) [reply modules/labs-react/combobox/spec/Combobox.spec.tsx]: That's fine. React Testing Library uses `act` internally which synchronously flushes rendering updates. This is why `get*` works almost always. I really like the queries in React Testing Library, but honestly the differences between `get*`, `find*` and `query*` are kind of a mess. I didn't understand the differences until I started contributing to the project...

File: modules/react/color-picker/lib/ColorInput.tsx
Comment (NicholasBoll) [reply modules/react/color-picker/lib/ColorInput.tsx:151]: `createComponent` doesn't take care of merging props, `mergeProps` does. Something like:

```tsx
const inputProps = mergeProps({
 onChange: handleChange,
  // other props you want
}, elemProps)
```

Then on `<CustomHexInput>`, you'd do something like:
```tsx
<ColorHexInput
  dir="ltr"
  // ...
  {...inputProps}
>
```

I'm not sure I'd worry about it though. I think it would be better to turn this into a compound component with behavior hooks that are exported which handle prop merging better.

If you wanted, this line could be simplified to:

```tsx
onChange?.(e)
```

Comment (alanbsmith) [reply modules/react/color-picker/lib/ColorInput.tsx:151]: Oh I was under the impression it was merging the css prop and other event handler functions. I’ll look back at the source again.

File: modules/codemod/lib/v5/renameInputRefs.ts
Comment (NicholasBoll) [inline modules/codemod/lib/v5/renameInputRefs.ts:67]: Glad the `getImportRenameMap` utility function was useful! Makes for simpler codemods

**Review by NicholasBoll — APPROVED** [review]

File: modules/codemod/lib/v5/renameInputRefs.ts
Comment (jamesfan) [reply modules/codemod/lib/v5/renameInputRefs.ts:67]: It was such a time-saver -- thanks for putting it together!

**Pull request merged**
